src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs

```diff
@@ -0,0 +1,417 @@
+                var kafkaMessage = @event.Message.ToKafkaMessage(ContentMode.Binary, formatter);
+
+                _kafkaProducer.Produce(@event.QueueName, kafkaMessage, ReportHandler);
+            }
+
+            _kafkaProducer.CommitTransaction();
+        }
+        catch (Exception ex)
+        {
+            try
+            {
+                _kafkaProducer.AbortTransaction();
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations($" ETL process: {Name}. Aborting Kafka transaction failed.", e);
+            }
+
+            throw new QueueLoadException(ex.Message, ex);
+        }
+
+        return count;
+    }
+
+    private int ProcessRabbitMq(IEnumerable<QueueWithMessages> items, BlittableJsonEventBinaryFormatter formatter)
+    {
+        int count = 0;
+
+        var messages = new List<RabbitMqMessageEvent>();
+
+        foreach (QueueWithMessages queueItem in items)
+        {
+            var newMessage = new RabbitMqMessageEvent() { Queue = queueItem.Name };
+
+            if (_queuesForDeclare.Any(x => x.QueueName == queueItem.Name) == false)
+            {
+                _queuesForDeclare.Add(new QueueDeclare() { QueueName = queueItem.Name });
+            }
+
+            foreach (var message in queueItem.Messages)
+            {
+                if (string.IsNullOrWhiteSpace(message.Options?.Exchange) == false)
+                {
+                    var queue = _queuesForDeclare.First(x => x.QueueName == queueItem.Name);
+
+                    if (queue.Exchanges.Any(x => x.ExchangeName == message.Options.ExchangeType) == false)
+                    {
+                        queue.Exchanges.Add(new ExchangeDeclare()
+                        {
+                            ExchangeName = message.Options?.Exchange,
+                            ExchangeType = message.Options?.ExchangeType ?? ExchangeType.Direct,
+                        });
+                    }
+                }
+
+                CloudEvent eventMessage = CreateEventMessage(message);
+                newMessage.Messages.Add(new RabbitMqMessage()
+                {
+                    Exchange = message.Options?.Exchange ?? "",
+                    ExchangeType = message.Options?.ExchangeType ?? ExchangeType.Direct,
+                    Message = eventMessage.ToAmqpMessage(ContentMode.Binary, formatter)
+                });
+            }
+
+            messages.Add(newMessage);
+        }
+
+        if (messages.Count == 0) return count;
+
+        if (_rabbitMqProducer == null)
+        {
+            _rabbitMqProducer = QueueHelper.CreateRabbitMqClient(Configuration.Connection, TransactionalId);
+        }
+
+        // todo: check application properties, _rabbitMqProducer.CreateBasicProperties()
+        // todo: try to manually populate properties
+
+        SetUpExchangesAndQueues(messages);
+        
+        foreach (var message in messages)
```

Comment:
`messages` name is confusing. When I looked at the code first time I was wondering why we open a transaction per message

Reply:
Ok, I will find better names